### PR TITLE
Mark BufferedChannelIterator.continuation @BeningDataRace to address …

### LINKED ISSUE #3834: Data race on `BufferedChannelIterator.continuation`
Our internal TSan tests found the following data race. I patched a fix locally to put `@Volatile` on `BufferedChannelIterator.continuation`, but I'm not sure if that's appropriate to upstream.

```
Details
==================
WARNING: ThreadSanitizer: data race (pid=9770)
  Write of size 4 at 0x0000cfe3a350 by thread T78:
    #0 kotlinx.coroutines.channels.BufferedChannel$BufferedChannelIterator.tryResumeHasNextOnClosedChannel()V BufferedChannel.kt:1697 
    #1 kotlinx.coroutines.channels.BufferedChannel.resumeWaiterOnClosedChannel(Lkotlinx/coroutines/Waiter;Z)V BufferedChannel.kt:2174 
    #2 kotlinx.coroutines.channels.BufferedChannel.resumeReceiverOnClosedChannel(Lkotlinx/coroutines/Waiter;)V BufferedChannel.kt:2161 
    #3 kotlinx.coroutines.channels.BufferedChannel.cancelSuspendedReceiveRequests(Lkotlinx/coroutines/channels/ChannelSegment;J)V BufferedChannel.kt:2154 
    #4 kotlinx.coroutines.channels.BufferedChannel.completeClose(J)Lkotlinx/coroutines/channels/ChannelSegment; BufferedChannel.kt:1931 
    #5 kotlinx.coroutines.channels.BufferedChannel.isClosed(JZ)Z BufferedChannel.kt:2210 
    #6 kotlinx.coroutines.channels.BufferedChannel.isClosedForSend0(J)Z BufferedChannel.kt:2185 
    #7 kotlinx.coroutines.channels.BufferedChannel.isClosedForSend()Z BufferedChannel.kt:2182 
    #8 kotlinx.coroutines.channels.BufferedChannel.completeCloseOrCancel()V BufferedChannel.kt:1903 
    #9 kotlinx.coroutines.channels.BufferedChannel.closeOrCancelImpl(Ljava/lang/Throwable;Z)Z BufferedChannel.kt:1796 
    #10 kotlinx.coroutines.channels.BufferedChannel.close(Ljava/lang/Throwable;)Z BufferedChannel.kt:1755 
    #11 kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Lkotlinx/coroutines/channels/SendChannel;Ljava/lang/Throwable;ILjava/lang/Object;)Z Channel.kt:98 
    #12 kotlinx.coroutines.channels.ProducerCoroutine.onCompleted(Lkotlin/Unit;)V Produce.kt:143 
    #13 kotlinx.coroutines.channels.ProducerCoroutine.onCompleted(Ljava/lang/Object;)V Produce.kt:136 
    #14 kotlinx.coroutines.AbstractCoroutine.onCompletionInternal(Ljava/lang/Object;)V AbstractCoroutine.kt:93 
    #15 kotlinx.coroutines.JobSupport.tryFinalizeSimpleState(Lkotlinx/coroutines/Incomplete;Ljava/lang/Object;)Z JobSupport.kt:296 
    #16 kotlinx.coroutines.JobSupport.tryMakeCompleting(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object; JobSupport.kt:860 
    #17 kotlinx.coroutines.JobSupport.makeCompletingOnce$third_party_kotlin_kotlinx_coroutines_kotlinx_coroutines_jvm(Ljava/lang/Object;)Ljava/lang/Object; JobSupport.kt:832 
    #18 kotlinx.coroutines.AbstractCoroutine.resumeWith(Ljava/lang/Object;)V AbstractCoroutine.kt:100 
    #19 kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(Ljava/lang/Object;)V ContinuationImpl.kt:46 
    #20 kotlinx.coroutines.DispatchedTask.run()V DispatchedTask.kt:108 
    #21 kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(Lkotlinx/coroutines/scheduling/Task;)V CoroutineScheduler.kt:584 
    #22 kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(Lkotlinx/coroutines/scheduling/Task;)V CoroutineScheduler.kt:793 
    #23 kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker()V CoroutineScheduler.kt:697 
    #24 kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run()V CoroutineScheduler.kt:684 
    #25 (Generated Stub) <null> 

  Previous read of size 4 at 0x0000cfe3a350 by thread T4:
    #0 kotlinx.coroutines.channels.BufferedChannel$BufferedChannelIterator.invokeOnCancellation(Lkotlinx/coroutines/internal/Segment;I)V BufferedChannel.kt:1647 
    #1 kotlinx.coroutines.channels.BufferedChannel.prepareReceiverForSuspension(Lkotlinx/coroutines/Waiter;Lkotlinx/coroutines/channels/ChannelSegment;I)V BufferedChannel.kt:718 
    #2 kotlinx.coroutines.channels.BufferedChannel.access$prepareReceiverForSuspension(Lkotlinx/coroutines/channels/BufferedChannel;Lkotlinx/coroutines/Waiter;Lkotlinx/coroutines/channels/ChannelSegment;I)V BufferedChannel.kt:37 
    #3 kotlinx.coroutines.channels.BufferedChannel$BufferedChannelIterator.hasNextOnNoWaiterSuspend(Lkotlinx/coroutines/channels/ChannelSegment;IJLkotlin/coroutines/Continuation;)Ljava/lang/Object; BufferedChannel.kt:3116 
    #4 kotlinx.coroutines.channels.BufferedChannel$BufferedChannelIterator.hasNext(Lkotlin/coroutines/Continuation;)Ljava/lang/Object; BufferedChannel.kt:1610 
    #5 kotlinx.coroutines.channels.ChannelsKt__Channels_commonKt.toList(Lkotlinx/coroutines/channels/ReceiveChannel;Lkotlin/coroutines/Continuation;)Ljava/lang/Object; Channels.common.kt:160 
    #6 kotlinx.coroutines.channels.ChannelsKt.toList(Lkotlinx/coroutines/channels/ReceiveChannel;Lkotlin/coroutines/Continuation;)Ljava/lang/Object; ??:1 
    #7 kotlinx.coroutines.channels.ChannelsKt__Channels_commonKt$toList$1.invokeSuspend(Ljava/lang/Object;)Ljava/lang/Object; Channels.common.kt 
    #8 kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(Ljava/lang/Object;)V ContinuationImpl.kt:33 
    #9 kotlinx.coroutines.DispatchedTask.run()V DispatchedTask.kt:108 
    #10 kotlinx.coroutines.EventLoopImplBase.processNextEvent()J EventLoop.common.kt:280 
    #11 kotlinx.coroutines.BlockingCoroutine.joinBlocking()Ljava/lang/Object; Builders.kt:85 
    #12 kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Lkotlin/coroutines/CoroutineContext;Lkotlin/jvm/functions/Function2;)Ljava/lang/Object; Builders.kt:59 
    #13 kotlinx.coroutines.BuildersKt.runBlocking(Lkotlin/coroutines/CoroutineContext;Lkotlin/jvm/functions/Function2;)Ljava/lang/Object; ??:1 
    #14 kotlinx.coroutines.TestBase.runTest(Lkotlin/jvm/functions/Function1;Ljava/util/List;Lkotlin/jvm/functions/Function2;)V TestBase.kt:224 
    #15 kotlinx.coroutines.TestBase.runTest$default(Lkotlinx/coroutines/TestBase;Lkotlin/jvm/functions/Function1;Ljava/util/List;Lkotlin/jvm/functions/Function2;ILjava/lang/Object;)V TestBase.kt:216 
    #16 kotlinx.coroutines.channels.TickerChannelCommonTest.testComplexOperator()V TickerChannelCommonTest.kt:97 
```

### PR DESCRIPTION
**Pull request by qwwdfsad**
…potential UB and OoTA on K/N

Fixes #3834

### CODE REVIEW

File: kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt [hunk 1]
```diff
@@ -1583,6 +1583,7 @@ internal open class BufferedChannel<E>(
          * continuation. The [tryResumeHasNext] and [tryResumeHasNextOnClosedChannel]
          * function resume this continuation when the [hasNext] invocation should complete.
          */
+        @BenignDataRace // See its uses for the explanation of the race
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt:1586]: I think we should explicitly list the races right next to the annotation. This way, if a new race is discovered, we won't look at the annotation and go, "oh yeah, we know about this race, it's benign, no worries."

File: kotlinx-coroutines-core/common/src/selects/Select.kt [hunk 2]
```diff
@@ -372,7 +372,12 @@ internal open class SelectImplementation<R>(
 
     /**
      * List of clauses waiting on this `select` instance.
+     *
+     * This property is the subject to bening data-race: concurrent cancellation might null-out this property
+     * while [trySelect] operation reads it and iterates over its content.
+     * A logical race is resolved by the consensus on [state] property.
      */
+    @BenignDataRace
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/common/src/selects/Select.kt:380]: I like the explanation here, for example.

File: kotlinx-coroutines-core/common/src/selects/Select.kt [hunk 3]
```diff
@@ -621,9 +627,11 @@ internal open class SelectImplementation<R>(
                         // try to resume the continuation.
                         this.internalResult = internalResult
                         if (cont.tryResume(onCancellation)) return TRY_SELECT_SUCCESSFUL
-                        // If the resumption failed, we need to clean
-                        // the [result] field to avoid memory leaks.
-                        this.internalResult = null
+                        /*
+                         * If the resumption failed, we need to clean the [result] field to avoid memory leaks.
+                         * This write is benignly races with the very same write in cancellation invoke() handler
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/common/src/selects/Select.kt:632]: And this one I don't like: to (ensure the correctness of)/(understand) this code, I don't think we need to know that there's a race.

A typo:
```suggestion
                         * This write benignly races with the very same write in cancellation invoke() handler
```

**Review by dkhalanskyjb — APPROVED** [review]
I think we could document this more reliably and intuitively, but other than that, everything looks good.

File: kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt (hunk 1 above)
Comment (qwwdfsad) [reply kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt]: Good idea, fixed

**Review by dkhalanskyjb — APPROVED** [review]

**Pull request merged**